pandas/core/reshape/reshape.py

```diff
@@ -1072,6 +1072,223 @@ def get_empty_frame(data) -> DataFrame:
+
+    >>> pd.from_dummies(d, dropped_first=["d", "e"])
+       C   col1    col2
+    0  1     a       b
+    1  2     b       a
+    2  3     d       e
+
+    >>> d = pd.DataFrame({"col1_a-a": [1, 0, 1], "col1_b-b": [0, 1, 0],
+    ...                   "col2-a_a": [0, 1, 0], "col2-b_b": [1, 0, 0],
+    ...                   "col2-c_c": [0, 0, 1]})
+
+    >>> pd.from_dummies(d, prefix_sep={"col1": "_", "col2": "-"})
+       col1  col2
+    0  a-a   b-b
+    1  b-b   a-a
+    2  a-a   c-c
+    """
+    from pandas.core.reshape.concat import concat
+
+    if data.isna().any().any():
+        raise ValueError(
+            f"Dummy DataFrame contains NA value in column: "
+            f"'{data.columns[data.isna().any().argmax()]}'"
+        )
+
+    if to_series:
+        return _from_dummies_1d(data, dropped_first)
+
+    data_to_decode: DataFrame
+    if columns is None:
+        columns = data.columns
+    elif not is_list_like(columns):
+        raise TypeError("Argument for parameter 'columns' must be list-like")
+    # index data with a list of all columns that are dummies
+    cat_columns = []
+    non_cat_columns = []
+    for col in columns:
+        if any(ps in col for ps in prefix_sep):
+            cat_columns.append(col)
+        else:
+            non_cat_columns.append(col)
+    data_to_decode = data[cat_columns].astype("boolean")
+    non_cat_data = data[non_cat_columns]
+
+    # get separator for each prefix and lists to slice data for each prefix
+    if isinstance(prefix_sep, dict):
+        variables_slice: dict[str, list] = {prefix: [] for prefix in prefix_sep}
+        for col in data_to_decode.columns:
+            for prefix in prefix_sep:
+                if prefix in col:
+                    variables_slice[prefix].append(col)
+    else:
+        sep_for_prefix = {}
+        variables_slice = {}
+        for col in data_to_decode.columns:
+            ps = [ps for ps in prefix_sep if ps in col][0]
+            prefix = col.split(ps)[0]
+            if prefix not in sep_for_prefix:
+                sep_for_prefix[prefix] = ps
+            if prefix not in variables_slice:
+                variables_slice[prefix] = [col]
+            else:
+                variables_slice[prefix].append(col)
+        prefix_sep = sep_for_prefix
+
+    # validate number of dropped_first
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})."
+            )
+            raise ValueError(len_msg)
+
+    if dropped_first:
+        if isinstance(dropped_first, dict):
+            check_len(dropped_first, "dropped_first")
+        elif isinstance(dropped_first, list):
+            check_len(dropped_first, "dropped_first")
+            dropped_first = dict(zip(variables_slice, dropped_first))
```

Comment:
How strict do we want to be with the the passed type: strict list only or allow all "workable" types?

Reply:
Made it strict for now